cpp/include/cudf/io/writers.hpp

```diff
@@ -207,6 +207,148 @@ class writer {
 
 }  // namespace parquet
 
+//! CSV format
+namespace csv {
+
+/**
+ * @brief Options for the CSV writer. 
+ * Also base class for `write_csv_args`
+ */
+struct writer_options {
+  
+  writer_options() = default;
+  writer_options(writer_options const&) = default;
+
+  virtual ~writer_options(void) = default;
```

Comment:
Does this really need a virtual destructor? I may be missing something as this is still WIP, but as is I don't see a need for inheritance.

Reply:
Let me first answer question# 1: "why virtual destructor?"

When inheriting `private`-ly polymorphism doesn't work _outside_ of the 2 classes that are involved in the `private` inheritance. However, polymorphism _does_ work inside member functions of the derived class. Snippet example:

```
class Base {
      virtual void foo(void);
};
class Derived: private Base {
  void foo(void) override;
};
```

Say now we have this function:
```
void bar(void) {
    std::unique_ptr<Base> ppx(new Base());
    std::unique_ptr<Base> ppy(new Derived());

    std::vector<std::unique_ptr<Base>> v(2);

    v[0] = std::move(ppx);
    v[1] = std::move(ppy);

    for(auto&& p: v)
      p->foo();
}
```
As a free-standing function this would not compile, so no need for a virtual destructor because the client code cannot see the polymorphism inside `bar()`.
_However_ `Derived::bar()`, i.e., `bar()`, with the same body as above, as a _member function_: well, _that_ would work just fine! Provided `Base` has a `virtual` destructor. Otherwise memory would leak. So, then this approach leaves the possibility of still taking advantage of polymorphism, as long as it's done inside `Derived`.

Question #2: Why using inheritance as opposed, I assume, aggregation (composition)?

I think I've answered this question above (https://github.com/rapidsai/cudf/pull/4484#discussion_r401938763), but to repeat:

Say, you use aggregation:
```
class Base {
      void foo(void);
};
class Derived{
      Base& get_base(void)
      {
         return base_;
      }
private:
      Base base_;
};
//...
int main(void)
{
   Derived d{...};
   d.get_base().foo();
}
```
With this setup, if I need to call `Base` memf's I need to call them via `get_base()`. I mean, imagine you have tens of such base memfs that you need to call from `Derived` objects in, say, 10,000 places. That's a lot of typing the additional `get_base()`, isn't it? (not worried about the additional time spent in the indirection because the compiler would optimize that). 

While with `private` inheritance solution, you just do:
```
class Base {
      virtual void foo(void);
};
class Derived: private Base {
  using Base::foo;
};
```
And, Booyah! the `foo` got injected inside `Derived` without any need for an indirection. More readable and saves a lot of typing.

Reply:
Also, inheritance was suggested in the issue discussion. See: https://github.com/rapidsai/cudf/issues/4342#issuecomment-606677618

Reply:
That's probably going to be painful for the cython layer...